pkg/operator/controller/instance_controller_test.go

```diff
@@ -0,0 +1,319 @@
+
+		// Check ingress resource not exist
+		assert.True(t, checkResourceNotExist(ingressID, &corev1.Service{}))
+	})
+
+	t.Run("Update ClusterLink deployment", func(t *testing.T) {
+		svc := &corev1.Service{}
+		// Update ingress resource to LoadBalancer type
+		getResource(types.NamespacedName{Name: ClusterLinkName, Namespace: controller.OperatorNameSpace}, &cl)
+		cl.Spec.Ingress.Type = clusterlink.IngressTypeLoadBalancer
+		Expect(k8sClient.Update(ctx, &cl)).Should(Succeed())
+		checkResourceCreated(ingressID, svc)
+		Expect(svc.Spec.Type).Should(Equal(corev1.ServiceTypeLoadBalancer))
+
+		// Update ingress resource to NodPport type
+		Expect(k8sClient.Delete(ctx, svc)).Should(Succeed())
+		getResource(types.NamespacedName{Name: ClusterLinkName, Namespace: controller.OperatorNameSpace}, &cl)
+		cl.Spec.Ingress.Type = clusterlink.IngressTypeNodePort
+		cl.Spec.Ingress.Port = 30444
+		Expect(k8sClient.Update(ctx, &cl)).Should(Succeed())
+		checkResourceCreated(ingressID, svc)
+		Expect(svc.Spec.Type).Should(Equal(corev1.ServiceTypeNodePort))
+		Expect(svc.Spec.Ports[0].NodePort).Should(Equal(int32(30444)))
+
+		// Update dataplane and controlpane
+		goReplicas := 2
+		loglevel := "debug"
+		containerRegistry := "quay.com"
+		imageTag := "v1.0.1"
+		cp := &appsv1.Deployment{}
+		dp := &appsv1.Deployment{}
+
+		getResource(cpID, cp)
+		Expect(k8sClient.Delete(ctx, cp)).Should(Succeed())
+		getResource(dpID, dp)
+		Expect(k8sClient.Delete(ctx, dp)).Should(Succeed())
+		getResource(types.NamespacedName{Name: ClusterLinkName, Namespace: controller.OperatorNameSpace}, &cl)
+
+		/// Update Spec
+		cl.Spec.DataPlane.Type = clusterlink.DataplaneTypeGo
+		cl.Spec.DataPlane.Replicas = goReplicas
+		cl.Spec.LogLevel = loglevel
+		cl.Spec.ImageTag = imageTag
+		cl.Spec.ContainerRegistry = containerRegistry
+		Expect(k8sClient.Update(ctx, &cl)).Should(Succeed())
+
+		/// Check controlplane
+		checkResourceCreated(cpID, cp)
+		cpImage := containerRegistry + "/" + controller.ControlPlaneName + ":" + imageTag
+		Expect(cp.Spec.Template.Spec.Containers[0].Image).Should(Equal(cpImage))
+		Expect(cp.Spec.Template.Spec.Containers[0].Args[1]).Should(Equal(loglevel))
+
+		/// Check dataplane
+		checkResourceCreated(dpID, dp)
+		goImage := containerRegistry + "/" + controller.GoDataPlaneName + ":" + imageTag
+		Expect(dp.Spec.Template.Spec.Containers[0].Image).Should(Equal(goImage))
+		Expect(*dp.Spec.Replicas).Should(Equal(int32(goReplicas)))
+		Expect(dp.Spec.Template.Spec.Containers[0].Args[1]).Should(Equal(loglevel))
+	})
+
+	t.Run("Delete ClusterLink deployment", func(t *testing.T) {
+		// Delete ClusterLink deployment
+		Expect(k8sClient.Delete(ctx, &cl)).Should(Succeed())
+
+		// Controlplane resources
+		for _, r := range cpResource {
+			checkResourceDeleted(cpID, r)
+		}
+
+		// Dataplane resources
+		for _, r := range dpResource {
+			checkResourceDeleted(dpID, r)
+		}
+
+		// Ingress resource
+		checkResourceDeleted(ingressID, &corev1.Service{})
+	})
```

Comment:
Can you please add a test for multiple clusterlink.Instance objects and their changes? Only the first should be in effect and when deleted the next in line should be realized.

Reply:
Done in the e2e test